Support local plantuml binary

puremourning opened the pull request:
Fixes #50 

This is just a very quick implementation, shariing in the hope that it's useful to people tracking #50. I have  need to run plantuml locally without the hassle of spinning up a server using docker. It looked easy to run `plantuml` directly, so I made this quick change.

There are some drawbacks to this:

1) it could better use `asyncio` and do the generations concurrently (switch to `asyncio.create_subprocess_shell` etc.)
2) the configuration is a little quirky in that it picks one of the two options if both supplied.
3) I didn't think too hard about error handling

disclaimer: I was "inspired" by https://github.com/mikitex70/plantuml-markdown/blob/master/plantuml_markdown/plantuml_markdown.py#L497

Commit:
Support local plantuml binary

A command can be provided along with some arguments to run it. This
allows for avoiding remote services and spinning up docker containers,
particularly useful for build systems or isolated hosts.

MikhailKravets commented:
Hi @puremourning! Thank you for the contribution! Sofar I've run CI checks and I'll review the PR in more details a bit later.

puremourning commented:
Fixed flake8 lints.

puremourning commented:
Fixed tests. I didn't write any new tests (yet); this is more of an RFC at this point.

MikhailKravets commented on docs/docs/getting_started/setup.md:
```diff
@@ -7,7 +7,11 @@
+        # For online service
         puml_url: https://www.plantuml.com/plantuml/
+        # Alternatively:
+        # puml_bin: /path/to/plantuml
+        # puml_args: [ '-timeout', '10', '-noerror', .... ]
```
What do you think about this configuration:

when no arguments to pass

```yaml
local: /path/to/plantuml
```

or expanded

```yaml
local:
    bin: /path/to/plantuml
    args: [ '-timeout', '10', '-noerror', .... ]
```

With **local** having higher priority than `puml_url`.

puremourning replied:
Seems OK. Alternatively we could just have a single 'local': make 'local' an array of strings, forming the initial command line to exec.

MikhailKravets replied:
Yes, that's even better.

```yaml
local: /path/to/plantuml
```

or

```yaml
local: ['path/to/puml', '-timeout', '10', '-noerror', ....]
```

puremourning replied:
looks tricky to make its type flexible with mkdocs API, but I'll check.

MikhailKravets commented on mkdocs_puml/puml.py:
```diff
@@ -28,6 +31,45 @@ def __str__(self):  # pragma: no cover
+
+    def translate(self, schemes: typing.Iterable[str]) -> typing.List[typing.Union[str, Fallback]]:
+        return [
+            self._generate(scheme) for scheme in schemes
```
It would be great to run the PlantUML binaries using a pool executor. This would allow multiple images to be rendered in parallel.

puremourning replied:
I think we can rely on asyncio to run them concurrently. should be simple.

MikhailKravets commented:
Hi @puremourning! I've added a few comments in code. Below is the list of what remains to do in the task

- [ ] Research / implement pool executor for diagrams generation
- [ ] Cover the code with tests
- [ ] Finalize configurations

I hope to have a window of time between late August and September to focus entirely on this project. So, I'll be able to catch up on the development. Really appreciate the work you’ve done so far!

puremourning commented:
- added asyncio to run generation concurrently
- switched to a single `puml_cmdline` config option which is a list. I kept the `puml_` prefix to match with `puml_url` , as I found that `local` was just a bit ambiguous. easy to change of course.

puremourning commented on README.md:
````diff
@@ -42,6 +42,14 @@ Bob -> Alice : hello
+```yaml
+plugins:
+  - plantuml:
+      puml_bin: plantuml
````
needs to be updated.

puremourning commented on docs/docs/getting_started/setup.md:
```diff
@@ -7,7 +7,11 @@
+        # For online service
         puml_url: https://www.plantuml.com/plantuml/
+        # Alternatively:
+        # puml_bin: /path/to/plantuml
```
needs updating

powellkjr commented:
Hi I also need this for an isolated environment. Is there anything I can do to help get this merged?
